frontend/src/components/worksheets/Worksheet/Worksheet.js

```diff
@@ -1121,26 +1121,30 @@ class Worksheet extends React.Component {
             if (this.hasEditPermission()) {
                 var editor = ace.edit('worksheet-editor');
                 if (saveChanges) {
-                    this.setState({
+                    // Use promise here to ensure the worksheet info will not be sent to the backend until the frontend state has finished updating
+                    const setAsyncState = (newState) =>
+                        new Promise((resolve) => this.setState(newState, resolve));
+                    setAsyncState({
```

Comment:
We could just do this instead (using `async` / `await` instead of callbacks):

```
await new Promise(e => this.setState({
                        ws: {
                            ...this.state.ws,
                            info: {
                                ...this.state.ws.info,
                                source: editor.getValue().split('\n'),
                            },
                        },
                    }, e));
```